### ISSUE TITLE
Hyrise does not compile with GCC 12

### ISSUE THREAD

**Issue by Bouncner** [hyrise/hyrise#2473]
As part of #2402 I tried to compile Hyrise with GCC 12. It fails, one reason is that std::unary is removed with C++20. That's easy to fix. The other issue is that the template definition of FrameOfReference segments is ambigious.

```
/hyrise/src/lib/storage/frame_of_reference_segment.hpp:37:78:   required from here
/usr/include/boost/hana/fwd/equal.hpp:69:12: error: ambiguous template instantiation for 'struct boost::hana::equal_impl<opossum::enum_constant_tag<opossum::EncodingType>, opossum::enum_constant_tag<opossum::EncodingType>, boost::hana::when<true> >'
   69 |     struct equal_impl : equal_impl<T, U, when<true>> { };
      |            ^~~~~~~~~~
In file included from /usr/include/boost/hana/concept/comparable.hpp:19,
                 from /usr/include/boost/hana/map.hpp:18,
                 from /hyrise/src/lib/all_type_variant.hpp:13:
/usr/include/boost/hana/equal.hpp:55:12: note: candidates are: 'template<class T, class U, bool condition> struct boost::hana::equal_impl<T, U, boost::hana::when<condition> > [with T = opossum::enum_constant_tag<opossum::EncodingType>; U = opossum::enum_constant_tag<opossum::EncodingType>; bool condition = true]'
   55 |     struct equal_impl<T, U, when<condition>> : default_ {
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/boost/hana/equal.hpp:104:12: note:                 'template<class C> struct boost::hana::equal_impl<C, C, boost::hana::when<(boost::hana::Constant<C>::value && boost::hana::Comparable<typename C::value_type>::value)> > [with C = opossum::enum_constant_tag<opossum::EncodingType>]'
  104 |     struct equal_impl<C, C, when<
      |            ^~~~~~~~~~~~~~~~~~~~~~
  105 |         hana::Constant<C>::value &&
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~
  106 |         Comparable<typename C::value_type>::value
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  107 |     >> {
      |     ~~
/usr/include/boost/hana/fwd/equal.hpp:69:12: error: invalid use of incomplete type 'struct boost::hana::equal_impl<opossum::enum_constant_tag<opossum::EncodingType>, opossum::enum_constant_tag<opossum::EncodingType>, boost::hana::when<true> >'
   69 |     struct equal_impl : equal_impl<T, U, when<true>> { };
```

This might be the issue: https://github.com/boostorg/hana/commit/4f5157bc425eac8e115d597142932fef2e535282
So it *might* be fixed with boost 1.75, which is not shipped with Ubuntu jammy.

**Comment by Bouncner** [comment]
This "horrible workaround" might be related: https://stackoverflow.com/questions/59250210/how-do-i-implement-equality-compare-operators-for-boost-hana-structs

**Comment by Bouncner** [comment]
Closing with merged #2611. I haven't tested GCC 12, but both GCC 11 and GCC 13 work fine. In case somebody ever needs explicitly GCC 12 and the mentioned issues persist, feel free to re-open this issue.
